src/core/lru.h

```diff
@@ -0,0 +1,181 @@
+// Copyright 2023, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+#pragma once
+
+#include <absl/container/node_hash_map.h>
+
+#include <optional>
+
+#include "base/logging.h"
+
+namespace dfly {
+
+template <typename T> class Lru {
+  struct Node {
+    const T* data_ptr;
+
+    uint32_t prev;
+    uint32_t next;
+
+    Node() : prev(0), next(0) {
+    }
+  };
+
+ public:
+  explicit Lru(size_t capacity) : head_(0), grow_size_(capacity) {
+    CHECK_GT(capacity, 1u);
+    node_arr_.resize(capacity);
+  }
+  ~Lru() {
+  }
+
+  std::optional<T> GetPrev(const T& data) const;
+  std::optional<T> GetLast() const;
+  enum class Position {
+    kHead,
+    kTail,
+  };
+  void Put(const T& data, Position position = Position::kHead);
+  void Remove(const T& data);
+
+  size_t Size() const {
+    return table_.size();
+  }
+
+ private:
+  void MoveToPosition(uint32_t index, Position position);
+
+  absl::node_hash_map<T, uint32_t> table_;  // map from item to index in node arr
+  std::vector<Node> node_arr_;
+  uint32_t head_;
+  size_t grow_size_;
+};
+
+template <typename T> std::optional<T> Lru<T>::GetPrev(const T& data) const {
+  auto it = table_.find(data);
+  if (it == table_.end()) {
+    return std::nullopt;
+  }
+  const auto& node = node_arr_[it->second];
+
+  DCHECK_EQ(node.data_ptr, &it->first);
+  const auto& node_prev = node_arr_[node.prev];
+
+  return *node_prev.data_ptr;
+}
+
+template <typename T> std::optional<T> Lru<T>::GetLast() const {
+  if (table_.size() == 0) {
+    return std::nullopt;
+  }
+  unsigned tail = node_arr_[head_].prev;
+  return *node_arr_[tail].data_ptr;
+}
+
+template <typename T> void Lru<T>::Put(const T& data, Position position) {
+  auto [it, inserted] = table_.emplace(data, table_.size());
+  if (inserted) {
+    if (it->second < node_arr_.size()) {
+      node_arr_.resize(node_arr_.size() + grow_size_);
```

Comment:
So I am not sure exponential will be good here.
In dash table we also not grow exponential but every time we want to grow we create one more segment which is constant size.
I am not so sure about this, I was thinking about 10K capacity (per thread). 
WDYT should be the initial capacity of the lru?

Reply:
The thing is, unlike dash table, each growth of a vector requires copying all items. And each copy gets more and more expensive. So on a 10m keys, 4-CPU server (not too big) we could have ~1k copies. That's not great.
Since we always push to the end, perhaps we should use a `deque` instead of a `vector` for this exact purpose?
But I think (not sure) that releasing entries at the end of the deque's bulk will free the bulk so that's also not ideal :|

I'd stick to exponential growth (I'm not sure why not still)